src/Models/LoginModel.php

```diff
@@ -67,6 +67,19 @@ public function recordLoginAttempt(
         $this->checkQueryReturn($return);
     }
 
+    /**
+     * Returns the previous login information for the user,
+     * useful to display to the user the last time the account
+     * was accessed.
+     */
+    public function previousLogin(User $user): ?Login
+    {
+        return $this->where('success', 1)
+            ->where('user_id', $user->id)
+            ->orderBy('id', 'desc')->get()
+            ->getRow(1, Login::class);
```

Comment:
I think it is better to use Model's methods as much as possible.

```php
            ->orderBy('id', 'desc')
            ->findAll(1, 1)[0];
```
@lonnieezell What do you think?

Reply:
I totally agree with you @kenjis , let's wait to here from @lonnieezell .

BTW, please can you shed more light or point me to where I can read up about creating test codes? I see there's a test folder, but I don't know how to use it.

Reply:
@sammyskills [Here ](https://forum.codeigniter.com/showthread.php?tid=81830)are some great resources for learning how to write unit tests.

Reply:
See https://github.com/codeigniter4/shield/blob/93f607f87e8690d897d8971829784364f42899b5/tests/Unit/UserTest.php#L70
You can copy the test method and modify for `previousLogin()`.

Reply:
You can run phpunit:
```console
$ vendor/bin/phpunit 
PHPUnit 9.5.21 #StandWithUkraine

Runtime:       PHP 8.0.21
Configuration: /Users/kenji/work/codeigniter/codeigniter-shield/phpunit.xml
Random Seed:   1658056341

...............................................................  63 / 214 ( 29%)
............................................................... 126 / 214 ( 58%)
............................................................... 189 / 214 ( 88%)
.........................                                       214 / 214 (100%)

Nexus\PHPUnit\Extension\Tachycardia identified these 3 slow tests:
+--------------------------------------------------------------------------+---------------+-------------+
| Test Case                                                                | Time Consumed | Time Limit  |
+--------------------------------------------------------------------------+---------------+-------------+
| Tests\\Controllers\\RegisterTest::testRegisterActionSuccess              | 00:00:00.99   | 00:00:00.50 |
| Tests\\Controllers\\RegisterTest::testRegisterRedirectsToActionIfDefined | 00:00:00.95   | 00:00:00.50 |
| Tests\\Unit\\DictionaryValidatorTest::testCheckTrueOnNotFound            | 00:00:00.78   | 00:00:00.50 |
+--------------------------------------------------------------------------+---------------+-------------+


Time: 00:13.253, Memory: 40.00 MB

OK (214 tests, 491 assertions)
```

Reply:
Thank you!

Reply:
I think Lonnie is still traveling, but I agree: better to use the Model native methods here.

Reply:
Done.